### PR TITLE
CERD-205 > Plot Graph Fixes > Improvements and bug fixes

### PR DESCRIPTION
**Pull request by drochgenius**
**WIP**

Hey @chrisunl ,

I am trying to use the plot-graph element in a problem, in the Problem Runner.
But it does not work as expected.

Would you please have a look?

Thanks

### CODE REVIEW

**Comment by drochgenius** [comment]
Hi @chrisunl ,

Looking at your latest changes, there is still work to do here.
I am unable to reduce the steps?
It doesn't seem to handle asymptotes propertly, as D3 crashes on NaN values. We'll need to handle this.

Have a good weekend

File: packages/plot-graph/src/components/plot-graph.ts [hunk 1]
```diff
@@ -165,20 +165,61 @@ export class PlotGraph extends ComponentBase<any> {
...
+            }
+        });
+
         // plot a line for each equation
         this.equationItems.forEach(equation => {
-            const equationXmin = parseInt(equation.getAttribute('equation-xmin'));
-            const equationXmax = parseInt(equation.getAttribute('equation-xmax'));
-            const equationYmin = parseInt(equation.getAttribute('equation-ymin'));
-            const equationYmax = parseInt(equation.getAttribute('equation-ymax'));
             const step = parseInt(equation.getAttribute('step'));
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Putting a step smaller than 1 doesn't work because of `parseInt`.
You should be using `parseFloat`.
Actually @chrisunl , any number that could potentially have decimals should be converted from string with `parseFloat`.

File: packages/component-base/src/components/coordinate-system.ts [hunk 2]
```diff
@@ -28,6 +29,15 @@ export class CoordinateSystem extends ComponentBase<HTMLElement[]> {
 
         this.value = elems;
     }
+
+    protected firstUpdated() {
+        const dGrid: string = this.getAttribute('draw-grid');
```
Comment (drochgenius) [inline packages/component-base/src/components/coordinate-system.ts]: Please use attribute reflection.

```
@property({ type: Boolean, reflect: true})
public grid: boolean;
```

File: packages/plot-graph/src/components/plot-graph.ts [hunk 3]
```diff
@@ -1,22 +1,24 @@
 import { html, TemplateResult, property, ComponentBase, CoordinateSystem } from '@hmh/component-base';
-import { line, curveMonotoneX } from 'd3-shape';
+import { line, curveBasis } from 'd3-shape';
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Why the change from `curveMonotoneX` to `curveBasis` ?

File: packages/plot-graph/src/components/plot-graph.ts [hunk 4]
```diff
@@ -165,58 +173,128 @@ export class PlotGraph extends ComponentBase<any> {
...
+        let xMin: number;
+        let xMax: number;
+        let yMin: number;
+        let yMax: number;
+        let xScale: ScaleLinear<number, number>;
+        let yScale: ScaleLinear<number, number>;
+        let xCrossPoint: number = 0;
+        let yCrossPoint: number = 0;
+
+        this.axes.forEach(axis => {
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: Add general comment to summarize what this `forEach` does.

File: packages/plot-graph/src/components/plot-graph.ts [hunk 5]
```diff
@@ -165,58 +173,128 @@ export class PlotGraph extends ComponentBase<any> {
...
+                yMin = parseFloat(axis.getAttribute('min'));
+                yMax = parseFloat(axis.getAttribute('max'));
+                yScale = this.scale(isDirectionX, yMin, yMax, /* istanbul ignore next */ height);
+                const crossingAttr: string = axis.getAttribute('crossing-point');
+                yCrossPoint = crossingAttr ? parseFloat(crossingAttr) : 0;
+            }
+        });
+
+        //draw the axes (assuming any have been added)
+        this.axes.forEach(axis => {
```
Comment (drochgenius) [inline packages/plot-graph/src/components/plot-graph.ts]: When I see this `forEach` and the one above, we're looping twice through the axes.
Would it be possible to do everything in one single forEach, so we loop only once?

**Comment by drochgenius** [comment]
Hi @chrisunl , a general comment here on performance.
Because it takes some time for D3 to load and draw the graph, we see that initially, the container has 0 height, until you append the SVG. 
This causes the page to reflow.
Since we know the graph size in advance, we should be able to display placeholder content, or at least set the container initial size vary quickly so it appears smoothing when the graph is actually rendered.
Ideally, we could display some loader image in place of the SVG.

This directive can be handy for such cases: https://polymer.github.io/lit-html/api/modules/_directives_until_.html

This said, we may want to save this enhancement for another ticket.

File: packages/component-base/src/components/coordinate-system.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { html, TemplateResult, ComponentBase } from './base.js';
  */
 export class CoordinateSystem extends ComponentBase<HTMLElement[]> {
     public value: HTMLElement[] = [];
+    private _drawGrid: boolean;
```
Comment (drochgenius) [inline packages/component-base/src/components/coordinate-system.ts]: This logic is wrong and overcomplicated.
Please leverage built-in behaviour of properties.

Comment (chrisunl) [reply packages/component-base/src/components/coordinate-system.ts]: changed

File: packages/plot-graph/src/components/plot-graph.ts (hunk 1 above)
Comment (chrisunl) [reply packages/plot-graph/src/components/plot-graph.ts]: yup

File: packages/plot-graph/src/components/plot-graph.ts (hunk 3 above)
Comment (chrisunl) [reply packages/plot-graph/src/components/plot-graph.ts]: Looks a little smoother
